### PR TITLE
[NNAPI QDQ] Add QDQTranspose op support

### PR DESCRIPTION
**Pull request by YUNQIUGUO**
**Description**: Describe your changes.

[NNAPI QDQ] Add QDQTranspose op support

*this pr's commits are based on this branch: #10483 can compare to see the changes

### CODE REVIEW

File: onnxruntime/test/providers/nnapi/nnapi_basic_test.cc [hunk 1]
```diff
@@ -359,6 +359,19 @@ TEST(NnapiExecutionProviderTest, TestQDQMul) {
...
+  // NNAPI use different rounding, which may cause ~1% difference in the result
+  RunQDQModelTest(BuildQDQTransposeTestCase<uint8_t /* InputType */,
+                                            uint8_t /* OutputType */>(
+                      {1, 3, 32, 32} /* input_shape */,
+                      {0, 3, 1, 2} /* perms */),
+                  "nnapi_qdq_test_graph_transpose",
+                  {
+                      true /* verify_entire_graph_use_ep */,
+                      1e-2f /* fp32_abs_err */,
+                  });
```
Comment (gwang-msft) [inline onnxruntime/test/providers/nnapi/nnapi_basic_test.cc:372]: Do you need this here? Transpose is just moving data around without calculation.

File: onnxruntime/test/optimizer/qdq_test_utils.h [hunk 2]
```diff
@@ -187,5 +187,27 @@ GetQDQTestCaseFn BuildBinaryOpTestCase(const std::vector<int64_t>& input_shape,
...
+GetQDQTestCaseFn BuildQDQTransposeTestCase(
+    const std::vector<int64_t>& input_shape,
+    const std::vector<int64_t>& perms) {
+    return [input_shape, perms](ModelTestBuilder& builder) {
+      auto* input_arg = builder.MakeInput<InputType>(input_shape, -128, 127);
+      auto* output_arg = builder.MakeOutput();
+
+      // add DQ
+      auto* dq_output = builder.MakeIntermediate();
+      builder.AddDequantizeLinearNode<InputType>(input_arg, .003f, 1, dq_output);
```
Comment (gwang-msft) [inline onnxruntime/test/optimizer/qdq_test_utils.h:201]: >1 [](http://example.com/codeflow?start=67&length=1)

Probably a different zero point than the int8 case, such as 
std::numeric_limits<InputType>::max() / 2

File: onnxruntime/test/optimizer/qdq_transformer_test.cc [hunk 3]
```diff
@@ -665,32 +665,18 @@ TEST(QDQTransformerTests, Gather) {
...
-      builder.AddQuantizeLinearNode<int8_t>(transpose_output, .003f, 1, output_arg);
-    };
-
+  auto test_case = [&](const std::vector<int64_t>& input_shape, const std::vector<int64_t>& perms) {
     auto check_graph = [&](InferenceSessionWrapper& session) {
       auto op_to_count = CountOpsInGraph(session.GetGraph());
       EXPECT_EQ(op_to_count["Transpose"], 1);
-      EXPECT_EQ(op_to_count["QuantizeLinear"], 0);
-      EXPECT_EQ(op_to_count["DequantizeLinear"], 0);
+      EXPECT_EQ(op_to_count["QuantizeLinear"], 1);
```
Comment (gwang-msft) [inline onnxruntime/test/optimizer/qdq_transformer_test.cc:672]: >1 [](http://example.com/codeflow?start=47&length=1)

Why is this changed 0 -> 1

File: onnxruntime/test/optimizer/qdq_test_utils.h [hunk 4]
```diff
@@ -187,5 +187,27 @@ GetQDQTestCaseFn BuildBinaryOpTestCase(const std::vector<int64_t>& input_shape,
...
+
+      // add Transpose
+      auto* transpose_output = builder.MakeOutput();  // transpose output is graph output
+      Node& transpose_node = builder.AddNode("Transpose", {dq_output}, {transpose_output});
+      transpose_node.AddAttribute("perm", perms);
+
+      // add Q
+      builder.AddQuantizeLinearNode<OutputType>(transpose_output, .003f, 1, output_arg);
```
Comment (gwang-msft) [inline onnxruntime/test/optimizer/qdq_test_utils.h:209]: Same here

File: onnxruntime/test/optimizer/qdq_transformer_test.cc (hunk 3 above)
Comment (YUNQIUGUO) [reply onnxruntime/test/optimizer/qdq_transformer_test.cc:672]: For this, if keep it as 0, the unit test QDQtransformerTest.Transpose will fail and show the actual Q/DQ op count to be 1. 

I am not 100% sure about this.. I figured maybe the qdq transformer does not do anything to dq-transpose-q?

File: onnxruntime/test/optimizer/qdq_test_utils.h [hunk 5]
```diff
@@ -187,5 +187,35 @@ GetQDQTestCaseFn BuildBinaryOpTestCase(const std::vector<int64_t>& input_shape,
...
+    int8_t q_zp = 1;
+#endif
+
+    // add DQ
+    auto* dq_output = builder.MakeIntermediate();
+    builder.AddDequantizeLinearNode<InputType>(input_arg, .003f, dq_zp, dq_output);
+
+    // add Transpose
+    auto* transpose_output = builder.MakeOutput();  // transpose output is graph output
+    Node& transpose_node = builder.AddNode("Transpose", {dq_output}, {transpose_output});
```
Comment (gwang-msft) [inline onnxruntime/test/optimizer/qdq_test_utils.h:213]: Transpose output will go to Q node, this is not a graph output, this could be the reason UT failes

Comment (YUNQIUGUO) [reply onnxruntime/test/optimizer/qdq_test_utils.h:213]: thanks so much for picking this. yes I just realized the reason is that I copied the transpose_nofusion qdqtransformer test cases, looks very similar.  but that one is the one where dq/q should not be dropped... fixed it;)

**Review by gwang-msft — APPROVED** [review]
:shipit:

File: onnxruntime/test/optimizer/qdq_test_utils.h [hunk 6]
```diff
@@ -187,5 +187,35 @@ GetQDQTestCaseFn BuildBinaryOpTestCase(const std::vector<int64_t>& input_shape,
...
+template <typename InputType, typename OutputType>
+GetQDQTestCaseFn BuildQDQTransposeTestCase(
+    const std::vector<int64_t>& input_shape,
+    const std::vector<int64_t>& perms) {
+  return [input_shape, perms](ModelTestBuilder& builder) {
+    auto* input_arg = builder.MakeInput<InputType>(input_shape, -128, 127);
+    auto* output_arg = builder.MakeOutput();
+
+#ifdef USE_NNAPI
+    uint8_t dq_zp = std::numeric_limits<InputType>::max() / 2;
```
Comment (gwang-msft) [inline onnxruntime/test/optimizer/qdq_test_utils.h:200]: >uint8_t [](http://example.com/codeflow?start=4&length=7)

InputType dq_zp 
OutputType q_zp


Same below

File: onnxruntime/test/optimizer/qdq_test_utils.h [hunk 7]
```diff
@@ -187,5 +187,35 @@ GetQDQTestCaseFn BuildBinaryOpTestCase(const std::vector<int64_t>& input_shape,
...
+    uint8_t q_zp = std::numeric_limits<OutputType>::max() / 2;
+#else
+    int8_t dq_zp = 1;
```
Comment (gwang-msft) [inline onnxruntime/test/optimizer/qdq_test_utils.h:203]: >1 [](http://example.com/codeflow?start=19&length=1)

1 is good for int8_t but not so good for uint8_t

File: onnxruntime/test/optimizer/qdq_test_utils.h [hunk 8]
```diff
@@ -187,5 +187,35 @@ GetQDQTestCaseFn BuildBinaryOpTestCase(const std::vector<int64_t>& input_shape,
...
+GetQDQTestCaseFn BuildQDQTransposeTestCase(
+    const std::vector<int64_t>& input_shape,
+    const std::vector<int64_t>& perms) {
+  return [input_shape, perms](ModelTestBuilder& builder) {
+    auto* input_arg = builder.MakeInput<InputType>(input_shape, -128, 127);
+    auto* output_arg = builder.MakeOutput();
+
+#ifdef USE_NNAPI
+    InputType dq_zp = std::numeric_limits<InputType>::max() / 2;
+    OutputType q_zp = std::numeric_limits<OutputType>::max() / 2;
```
Comment (gwang-msft) [inline onnxruntime/test/optimizer/qdq_test_utils.h:201]: You probably only need this part, not need for separated definitions for NNAPI and non-NNAPI

**Review by gwang-msft — APPROVED** [review]
